internal/pkg/db/mongo/data.go

```diff
@@ -55,14 +55,16 @@ func (mc MongoClient) AddEvent(e correlation.Event) (string, error) {
 	// Add the readings
 	if len(e.Readings) > 0 {
 		var ui []interface{}
+		timestamp := db.MakeTimestamp()
 		for i, reading := range e.Readings {
 			var r models.Reading
 			id, err := r.FromContract(reading)
 			if err != nil {
 				return "", err
 			}
 
-			r.TimestampForAdd()
+			r.Created = timestamp
```

Comment:
The Redis implementation checks to see if the value is provided before generating a timestamp and setting it. Should we be doing that here for consistency? See the code block here:

https://github.com/edgexfoundry/edgex-go/blob/f1941eeb15de8730bc21d10e7aa8954d0a8f4e76/internal/pkg/db/redis/data.go#L1035-L1037

Reply:
The mongo implementation generate an event's timestamp after readings is set.So no need to check it.
https://github.com/edgexfoundry/edgex-go/blob/7ca3df59c01c7f56d3c83a840c5995cdb7718c20/internal/pkg/db/mongo/data.go#L80-L87

Reply:
I apologize, I referred to different entities in my last comment and even confused myself when I went back to look at it. Allow me to start over and provide concrete examples for clarification.

When an event is created via the HTTP API, the call can optionally contain a `created` field. If it is not provided in the HTTP body or is `0` the `json.Marshal` function will create an `event` with the `created` field of `0`. When the Mongo or Redis gets passed an unmarshalled `event` the `AddEvent` logic for each implementation is handling the `created` field differently, resulting in an inconsistency.

The Mongo implementation is calling this function right before persisting to the database:
https://github.com/edgexfoundry/edgex-go/blob/eec8e1f3ce6d9a9c8c1e1098fbe9df49c355ee97/internal/pkg/db/mongo/models/event.go#L97-L100

Which is always setting the `created` value to the current time in milliseconds regardless if the original content has already been provided with a `created` value.

On the other hand, the Redis implementation is performing this check and only updating the `created` field if the original value is `0`, which means the caller either did not provide it or it was explicitly set to `0`. This logic is calling the same method to update the created time, but it is doing it only if the value was not set:
https://github.com/edgexfoundry/edgex-go/blob/f1941eeb15de8730bc21d10e7aa8954d0a8f4e76/internal/pkg/db/redis/data.go#L1035-L1037

I was able to verify this by starting `core-data` using a Mongo database and provided this request to create an event, notice that the `created` field is provided:

```json
{  
   "origin":1471806386919,
   "device":"cartachometer",
   "created": 62388,
   "readings":[  
      {  
        "origin":1471806386919,
		"device":"cartachometer",
        "name":"luminousIntensity",
        "value":"40"
      }
   ]
}
```
And then looked up the event with the ID that was returned from the creation API and got, notice the original `created` value was ignored and over written:
```json
{
    "id": "1d60e893-3021-4948-8e86-502c8638a022",
    "device": "cartachometer",
    "created": 1579877020172,
    "modified": 1579877020172,
    "origin": 1471806386919,
    "readings": [
        {
            "id": "ef33dacc-bd76-44a6-ad49-078598b19f2d",
            "created": 1579877020169,
            "origin": 1471806386919,
            "modified": 1579877020169,
            "device": "cartachometer",
            "name": "luminousIntensity",
            "value": "40"
        }
    ]
}
```

I then did the same with `core-data` using Redis. My creation request was:
```json
{  
   "origin":1471806386919,
   "device":"cartachometer",
   "created": 62388,
   "readings":[  
      {  
        "origin":1471806386919,
		"device":"cartachometer",
        "name":"luminousIntensity",
        "value":"40"
      }
   ]
}
```

And my lookup for the newly created event response was, notice that the `created` value provided was persisted:
```json
{
    "id": "d20eea1d-efcf-491f-98b9-15f8409ece51",
    "device": "cartachometer",
    "created": 62388,
    "origin": 1471806386919,
    "readings": [
        {
            "id": "37feee2c-fb45-4990-bde8-ecf76533db79",
            "created": 62388,
            "origin": 1471806386919,
            "device": "cartachometer",
            "name": "luminousIntensity",
            "value": "40"
        }
    ]
}
```

You can see that the data for the `created` field that was persisted in the database differs depending on which database implementation you are using.

So my suggestion is to add logic to the Mongo implementation so that it behaves like the Redis implementation with regards to adding an event. I am assuming that the Redis implementation is correct since it makes sense to let the user provide a `created` value if they are storing data that has been created previously. For example, if the system is down and data gets backed up, we should honor the original created times. That is just my thought.